netket/vqs/mc/mc_state/state.py

```diff
@@ -827,8 +827,8 @@ def deserialize_MCState(vstate, state_dict):
     new_vstate.reset()
 
     new_vstate.variables = jax.tree_util.tree_map(
-        jnp.asarray,
-        serialization.from_state_dict(vstate.variables, state_dict["variables"]),
+        lambda x, y: jax.lax.with_sharding_constraint(jnp.asarray(y), x.sharding),
+        vstate.variables, serialization.from_state_dict(vstate.variables, state_dict["variables"]),
```

Comment:
please split this into 3 'lines', one doing the jnp.asarray, the following doing the sharding constraint, the last assigning to variables.

Also, please run the `jax.lax.with_sharding_constraint` only if `config.netket_experimental_sharding` is true